### ISSUE TITLE
Backtrace-tracking feature for debugging OSAL mutexes

### ISSUE THREAD

**Issue by skliper** [nasa/osal#64]
Buggy code that takes a mutex but forgets to release it can be very difficult to debug - when the code is stopped in the debugger, it is clear that the thread is waiting on the mutex, but evidence of the real problem is long gone and it is impossible to tell which thread took the mutex but did not give it up.

glibc has a {{{backtrace()}}} function which can be extremely useful in debugging these situations.

This will add an enhancement to the "posix-ng" build to __optionally__ obtain and store a backtrace with each successful "take" of a mutex.  

If at some point in the future another thread gets locked trying to obtain the mutex, the backtrace will point exactly to the code that obtained the mutex without releasing it.

Obviously this has some performance/memory implications (although not much) so it will not be enabled by default.  This would be something the developer could enable when actually trying to debug mutex release issues.

**Comment by skliper** [comment]
_Imported from trac issue 41.  Created by jphickey on 2015-05-08T16:06:45, last modified: 2019-07-26T15:13:27_

**Comment by skliper** [comment]
_Trac comment by glimes on 2015-08-25 11:52:07:_

Cool idea. Hope tail-call optimizations don't hide
all the goodies from us.

Entirely encapsulated in an #ifdef.

Accept.

**Comment by skliper** [comment]
_Trac comment by acudmore on 2015-08-25 12:49:48:_

I like it, as long as it can be completely turned off ( which it sounds like it can )

Is there any other place where a backtrace might be used? ( performance monitor, exceptions, etc? )
Perhaps a general API to access this might be useful.

**Comment by skliper** [comment]
_Trac comment by jphickey on 2019-04-12 09:30:23:_

This just needs a quick CCB discussion and update before the next OSAL release.

The "posix-ng" already has this feature implemented and it is wrapped in an `#ifdef` that must be explicitly turned on if you want it.  It relies on a function `backtrace()` which is glibc-specific (technically not posix) to do this.

It may or may not be possible to implement this in other systems - depends if they offer something like `backtrace()` or not.  That being said, the information provided by this feature can be extremely valuable during debugging of mutex-related problems, so I'd like to keep the hooks in place at least.

A proposed compromise:  Add a user-specified hook/callback function for mutex (and maybe semaphore) give/take events.  This could be a compile-time option.  With this, the PSP could implement the backtrace, if desired.  This way the OSAL stays standards-based, and only the PSP would call the implementation-specific functions.

**Comment by skliper** [comment]
_Trac comment by jphickey on 2019-06-11 14:12:05:_

Commit [changeset:7190406] on branch `trac-41-remove-backtrace-hook` cleans this up per CCB decision.

Ready for review.

**Comment by skliper** [comment]
_Trac comment by jhageman on 2019-06-13 11:28:13:_

CCB 6/12/19 - CCB code reviewed and approved

**Comment by skliper** [comment]
_Trac comment by jphickey on 2019-06-25 11:54:45:_

Merged into `ic-ccb-20190612`

**Comment by skliper** [comment]
_Trac comment by jhageman on 2019-07-10 13:23:34:_

CCB 07/10/2019 - Approved for merge to dev

**Comment by skliper** [comment]
_Trac comment by jphickey on 2019-07-10 13:48:08:_

Now in the "development" branch

**Comment by skliper** [comment]
_Trac comment by jhageman on 2019-07-26 15:13:27:_

Milestone renamed
